### ISSUE TITLE
nuke RSS & custom db and replace with GraphQL requests

### ISSUE THREAD

**Issue by commonuserlol** [mob-sakai/unity-changeset#16]
NOTE: Your issue may already be reported! Please search on the [issue tracker](../) before creating one.

**Is your feature request related to a problem? Please describe.**
Using custom database requires extra steps to update it and looks not really good IMO, RSS reports only 16 latest editors.

**Describe the solution you'd like**
Instead, you can use GraphQL requests to Unity API, [feel free to borrow from here](https://github.com/commonuserlol/libil2cpp-parser/blob/main/src/downloader.rs) or if you don't know rust, intercept request using devtools with payload when you switch tabs with editor versions

**Comment by mob-sakai** [comment]
I did not know until now that unity.com uses GraphQL...
Thanks a lot! 👍

**Comment by mob-sakai** [comment]
Release API specification: https://services.docs.unity.com/release/v1/

Edit: This is REST version

**Comment by mob-sakai** [comment]
It is working well so far with RSS and a custom DB, but I plan to move to GraphQL in the near future!

**Comment by mob-sakai** [comment]
> https://services.docs.unity.com/release/v1/
> v1
> January 19, 2023: Initial Release

Wow

**Comment by webbertakken** [comment]
I had definitely also missed the announcement for that.

By the amount of endpoints and revisions it has (just the one), looks like it was a tough nut to crack in a large company like Unity Technologies and may have taken longer to build than expected causing focus to shift away from it again. Just speculation of course.

Edit: Also can confirm that this repo has been very stable since the beginning.

**Comment by commonuserlol** [comment]
> Release API specification: https://services.docs.unity.com/release/v1/
> 
> Edit: This is REST version

I dunno what would be better: GraphQL or REST. However REST looks a bit cleaner than graphql one (at top comment there's my solution to parse it's response)., and probably faster (graphql = 9 requests for unity 5~6000, rest 1 request I guess?).

**Comment by webbertakken** [comment]
The REST one definitely looks a lot cleaner.

**Comment by mob-sakai** [comment]
I did some testing with [Altair GraphQL Client](https://chromewebstore.google.com/detail/altair-graphql-client/flnheeellpciglgpaodhkhmapeljopja). 

- getUnityReleases
  - limit was up to 1000.
  - You can use `[LTS, TECH, ALPHA, BETA]` as stream.
  - version is forward matching, but you can get all versions by specifying `”. "` to get all versions.
  - If you query with `stream=[], version=”."` will give you all versions. It takes a long time the first time, but after that it is fast because it is cached.
- getUnityReleaseMajorVersions
  - I expected to get all major versions, but only “2021”, “2022”, and “6000” were returned.

The final resulting query:
```
query {
  getUnityReleases(
    version: "."
    skip: 0
    limit: 1000
    stream: []
  )
  {
    totalCount
    edges {
      node {
        version
        shortRevision
      }
    }
    pageInfo {
      hasNextPage
    }
  }
}
```

**Comment by mob-sakai** [comment]
The GraphQL version seems much faster and easier to implement.

- You can get the change set without referring to RSS or release documents.
- The request header was unnecessary.
- In general usage scenarios, 1000 of the most recent releases are sufficient.
- It is also possible to acquire all cases depending on the pagination.

**Comment by commonuserlol** [comment]
> * I expected to get all major versions, but only “2021”, “2022”, and “6000” were returned.

That's probably due limit? Since there's almost 1300 unity releases, I did request for each version (5, 2017..2023, 6000).

Also abt changeset - in both graphql and rest you just have to parse `unityHubDeepLink` (and probably it have direct download URLs for each platform? In preview I see only windows thought)

**Comment by mob-sakai** [comment]
Well, it is `getUnityReleaseMajorVersions` query that returns "2021", "2022", and "6000".
```
{
  getUnityReleaseMajorVersions(
    stream: []
    platform: []
    architecture: []
    entitlements: []
  ) {
    version
    latestUnityRelease {
      version
      productName
    }
  }
}
```

`getUnityReleases` query can specify versions (="5", "2017", ... , "2023", "6000"), but I do not want to update the code if values greater than 6000 are added in the future...
Therefore, I was looking for a query to get `["5", "2017", ... , "2023", "6000"]`.
But it does not seem to be there.

**Comment by commonuserlol** [comment]
It returns 2021, 2022 and 6000 since it's officially supported (and LTS). others are outdated so they don't list it. I don't actually know how to deal with future updates (e.g. they'll make 7000). I'mma look again today, maybe I miss something obvious on their website or graphql..

**Comment by commonuserlol** [comment]
Looks like it's hardcoded. I found JS with queries but nothing abt querying versions..

**Comment by mob-sakai** [comment]
Thank you.
I tested the GraphQL version of the implementation.  
https://github.com/mob-sakai/unity-changeset/blob/develop/src/unityGraphQL.ts

It seems to be working well. 👍  
https://github.com/mob-sakai/unity-changeset/actions/runs/11765701407

**Comment by mob-sakai** [comment]
Released as version 2.3.0. 🎉 
The custom DB will be left as is for now as an alternative.
